### ISSUE TITLE
Gitlab repository: support both the project ID and the escaped full path

### ISSUE THREAD

**Issue by aetertinas9** [kubernetes-sigs/cluster-api#10605]
### What steps did you take and what happened?

https://github.com/kubernetes-sigs/cluster-api/pull/9481

This is a bug that occurred in the above link and has not been resolved for a long time. However, it needs to be included in the release version, so I am opening a new issue.

### What did you expect to happen?

Using GitLab as per the [docs](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html?highlight=gitlab#provider-repositories), it supports both the project ID and the escaped full path.

### Cluster API version

release version

### Kubernetes version

v1.28.4

### Anything else you would like to add?

Thanks :)

### Label(s) to be applied

/kind bug
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by aetertinas9** [comment]
sorry it's my first PR

**Comment by aetertinas9** [comment]
@fabriziopandini 
That's correct. I also opened this issue because of the same problem with not being able to use the GitLab URL properly. Instead of using the url.Path, this supports both the escaped full path and project ID as per the official documentation.

**Comment by fabriziopandini** [comment]
/priority backlog

I think those kind of errors boils up to the fact that we don't have a good test coverage for all the supported use cases.
